### PR TITLE
Daniel/simple stats explorer

### PR DESCRIPTION
**Pull request by dandanlen**
A few things to note from this pull request: 

1. The `/explore` endpoint now returns basic stats on the requested column. For example: 

```
curl -k -X POST -H "Content-Type: application/json" \
   http://localhost:5000/explore  \
   -d '{"ApiKey":"SFMyNTY.g3QAAAACZAAEZGF0YW0AAAAkMTUxNjJiZWYtNWE2MS00NGNhLWFiZmUtOWU1MGFiNGIxM2M4ZAAGc2lnbmVkbgYAlH8NpW8B.byOGmraal0gWNKa_g6aXgArfff2nl34Tm-hJL43sOIw", \
"TableName":"loans", \
"ColumnName":"amount", \
"DataSourceName": "gda_banking"}'
```

Gives:
```javascript
{
  "status":"complete",
  "metrics":[
    {"Name":"Min","Type":"real","Value":36780},
    {"Name":"Max","Type":"real","Value":494767},
    {"Name":"Count","Type":"integer","Value":825},
    {"Name":"CountNoise","Type":"real","Value":1}
  ],
  "id":"443de6d2-f795-4e72-b43f-b0fc17ecad58"
}
```

I created a basic response type `ExploreResult` for this - I'm open to suggestions on this. (`ExplorationResult` may be better. Also, not sure about the `Metrics` class contained in it.)

Btw. Right now the json representation is not very consistent between UpperCase / lowercase, not quite sure why the "status", "metrics", "id" are all lower case...

2. The `NumericExplorer` object yields results asynchronously and for now I just write those status updates to the console. What we really want to do in the controller is more like: 
- Create the `ColumnExplorer` object
- Retrieve it's ID and store the associated Explorer object somewhere. 
- Kick off the actual Exploration `Task`
- Return the ID in the Http response

- We then need another endpoint like /results/{exploration_id} that can be polled for the `ExploreResult`
- When polled, the latest status can be retrieved using the Guid for the `ColumnExplorer` instance. 

This way the api is as similar as possible to the Aircloak api. 

I will create an issue for the 2nd point above. 

Next step for me: create some tests.

### CODE REVIEW

File: src/explorer.api/Models/ExploreResult.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    using System.Text.Json.Serialization;
+    using Aircloak.JsonApi.ResponseTypes;
+
+    internal class ExploreResult
+    {
+        public ExploreResult(Guid explorationId, string status)
+        {
+            Id = explorationId;
+            Status = status;
+            Metrics = new List<Metric>();
```
Comment (AndreiBozantan) [inline src/explorer.api/Models/ExploreResult.cs]: In order to avoid allocation of an object which will get unused, I think that we can use this instead,: `Metrics = Array.Empty<Metric>();`

File: src/explorer.api/Models/NumericOverview.cs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+                },
+                new ExploreResult.Metric("CountNoise")
+                {
+                    MetricType = AircloakType.Real,
+                    MetricValue = stats.CountNoise,
+                },
+            });
+        }
+
+        private class QuerySpec : IQuerySpec<ResultRow>
```
Comment (AndreiBozantan) [inline src/explorer.api/Models/NumericOverview.cs]: Is this class related just to the NumericOverview? At a first glance it seems to be somewhat more general.

File: src/explorer.api/Models/NumericOverview.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
+﻿namespace Explorer
```
Comment (AndreiBozantan) [inline src/explorer.api/Models/NumericOverview.cs]: I am not sure that this class and its base class (ColumnExplorer) belong under the Models folder.

File: tests/explorer.api.tests/ExploreTest.cs [hunk 4]
```diff
@@ -13,7 +14,7 @@ public sealed class ExploreTest
     {
         private const string apiKey = "SFMyNTY.g3QAAAACZAAEZGF0YW0AAAAkMTUxNjJiZWYtNWE2MS00NGNhLWFiZmUtOWU1MGFiNGIxM2M4ZAAGc2lnbmVkbgYAlH8NpW8B.byOGmraal0gWNKa_g6aXgArfff2nl34Tm-hJL43sOIw";
```
Comment (AndreiBozantan) [inline tests/explorer.api.tests/ExploreTest.cs]: Is it safe to store the API key here?

File: src/explorer.api/Models/NumericOverview.cs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        }
+
+        private class ResultRow : IJsonArrayConvertible
+        {
+#pragma warning disable RCS1074 // Remove redundant constructor
+            // Need this for JsonArrayConverter constraint
+            public ResultRow()
+            {
+            }
+#pragma warning restore RCS1074 // Remove redundant constructor
```
[the remark this answers is not in the document]
Comment (dandanlen) [reply src/explorer.api/Models/NumericOverview.cs]: You're correct, I'll remove it. 

I think I initially had another constructor here so I had to explicitly define a default paramaterless constructor to satisfy a type constraint.

File: src/explorer.api/Models/NumericOverview.cs [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+            internal double? Min { get; set; }
+
```
[the remark this answers is not in the document]
Comment (dandanlen) [reply src/explorer.api/Models/NumericOverview.cs]: We will eventually want to treat it as a number (to determine range bounds for example) so maybe decimal is better here. 

There are actually only two types of number that can be returned by Aircloak: `integer` or `real` so another possibility is to have two classes, one that uses int64 and another that uses doubles. <-- I think this might actually be a cleaner solution than the above.

What do you think?

File: src/explorer.api/Models/NumericOverview.cs [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        internal async IAsyncEnumerable<ExploreResult> Explore()
+        {
+            var queryParams = new QuerySpec(ExploreParams);
+
+            yield return new ExploreResult(ExplorationGuid, status: "waiting");
+
+            var queryResult = await ApiSession.Query<ResultRow>(
+                ExploreParams.DataSourceName,
+                queryParams.QueryStatement,
+                TimeSpan.FromMinutes(2));
```
[the remark this answers is not in the document]
Comment (dandanlen) [reply src/explorer.api/Models/NumericOverview.cs]: You're right, I realise I have been scattering some arbitrary constants around... All of this will need to be gathered in configuration or environment variables but I'm not quite sure the best way to do this in c#/ASP.net so I have been delaying it... 

Other things to go in config :
- the aircloak server root url
- the api key
- anything else?

Good point about the cancel query endpoint! I thought it seemed sensible to set a timeout but to better mirror the functionality of Aircloak a `/cancel` endpoint might be better.

File: src/explorer.api/Models/ExploreResult.cs (hunk 1 above)
Comment (dandanlen) [reply src/explorer.api/Models/ExploreResult.cs]: Thanks for the tip, I didn't know this!

File: src/explorer.api/Models/NumericOverview.cs (hunk 2 above)
Comment (dandanlen) [reply src/explorer.api/Models/NumericOverview.cs]: Yes, this class defines the query specifically for the NumericOverview so i decided to define it within that class.

In retrospect it could be used elsewhere too so maybe it would be better to extract it (and the `ResultRow` class) into a `Explorer.Queries` namespace or something like that...

I wasn't sure how best to organise this so am open to suggestions.

File: src/explorer.api/Models/NumericOverview.cs (hunk 3 above)
Comment (dandanlen) [reply src/explorer.api/Models/NumericOverview.cs]: Ok, I'm not sure I understand the asp.net definitions of models vs controllers vs everything else. Better in the top-level folder then? 

What kinds of things should go in the Models folder? Only objects that are actually sent to / received by the api?

File: tests/explorer.api.tests/ExploreTest.cs (hunk 4 above)
Comment (dandanlen) [reply tests/explorer.api.tests/ExploreTest.cs]: Yeah totally safe, it'll never get lost. 😆 

I suppose best practice here is to pass it in as an environment variable or have a local config file that is never checked in to source control. However then I can't see how we would get the api key into Travis CI for testing..

File: src/explorer.api/Models/NumericOverview.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (AndreiBozantan) [reply src/explorer.api/Models/NumericOverview.cs]: Yes, I believe that it would be better to have two different classes, one that uses int64 and the other using double.

File: src/explorer.api/Models/NumericOverview.cs (hunk 3 above)
Comment (AndreiBozantan) [reply src/explorer.api/Models/NumericOverview.cs]: I am not sure also, I didn't use ASP.NET core before. But I think that the models should contain just the objects sent to/received by the API, as you said; that is simple classes that usually contain just fields and don't have too much additional functionality.

File: tests/explorer.api.tests/ExploreTest.cs (hunk 4 above)
Comment (AndreiBozantan) [reply tests/explorer.api.tests/ExploreTest.cs]: I also don't know how to get API keys into Travis CI without involving git. It seems that there is some documentation here https://docs.travis-ci.com/user/private-dependencies/ but I am not sure if it is worth investigating it. @sebastian  what do you think about this?

File: src/explorer.api/Models/NumericOverview.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (dandanlen) [reply src/explorer.api/Models/NumericOverview.cs]: Ok i've done this in the latest push, broke it apart into two classes and extracted the query definition into its own namespace.

File: src/explorer.api/Models/NumericOverview.cs (hunk 3 above)
Comment (dandanlen) [reply src/explorer.api/Models/NumericOverview.cs]: That makes sense. I've reorganised the file structure in the latest push.

File: tests/explorer.api.tests/ExploreTest.cs (hunk 4 above)
Comment (sebastian) [reply tests/explorer.api.tests/ExploreTest.cs]: Definitely not safe to store API keys here.
This key has already been compromised and should be immediately revoked.
You can set environmental variables for Travis to use on Travis itself.

Comment (sebastian) [reply tests/explorer.api.tests/ExploreTest.cs]: We do not need the private dependencies features. It's for testing private repositories whereas this is a public one.

Please check if you have access to the Travis Explorer repo config when logging in to travis using your github credentials. If you do, then you can set the environmental variable there. If for some reason it does not work, let me know, and pass me a key via an out of bounds mechanism that I can add for you instead.

Comment (sebastian) [reply tests/explorer.api.tests/ExploreTest.cs]: This is a URL you can use to check if you have access. It's also where the setting would be changed: https://travis-ci.org/diffix/explorer/settings

Comment (dandanlen) [reply tests/explorer.api.tests/ExploreTest.cs]: Thanks for the link, didn't know we had that! I've generated a new key and added it to the environment, will test it now.

Comment (dandanlen) [reply tests/explorer.api.tests/ExploreTest.cs]: Ok all set up now. 

@AndreiBozantan if you want to run the tests locally you need an API key added to your environment (https://attack.aircloak.com/api_tokens) under AIRCLOAK_API_KEY. You probably also have to rebuild the development container. 

Fyi. The link should be: https://travis-ci.org/diffix/explorer/settings
(note .com instead of .org, apparently .org is being phased out).
